### DISCUSSION TITLE
history change fired twice on initial page load which leads to wrong analytics result

### DISCUSSION THREAD

**Discussion by macrozone** [vercel/next.js#9557] [Help]
# Bug report

## Describe the bug

nextjs calls the history api twice with the same values on initial page load. This gives wrong analytics results if you define a "history change event" trigger in google tag manager:

<img width="1094" alt="Bildschirmfoto 2019-11-27 um 17 56 06" src="https://user-images.githubusercontent.com/1972353/69743878-78788e80-113f-11ea-86e1-34c59aea183f.png">
<img width="897" alt="Bildschirmfoto 2019-11-27 um 17 57 51" src="https://user-images.githubusercontent.com/1972353/69743893-7d3d4280-113f-11ea-8647-9989465d63ec.png">

## To Reproduce

1. add google tag manager to nextjs. in _document.jsx add something like:

```
 <Head>
          {GOOGLE_TAG_MANAGER_ID ? (
            <script
              dangerouslySetInnerHTML={{
                __html: `(function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':
          new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],
          j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src=
          'https://www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f);
        })(window,document,'script','dataLayer','${GOOGLE_TAG_MANAGER_ID}');`,
              }}
            />
          ) : null}
        </Head>
```

2. configure google tag manager to track history change events. refer to the google tag manager documentation
3. set your google tag manager into preview mode
4. this should add a debug bar to your application where you can see the events that are sent. You will notice that on initial page load the event is sent twice.

## Expected behavior

nextjs should not call the history api with the same arguments twice.

## Screenshots

see above

## System information

- OS: [e.g. macOS, Windows]
- Browser (if applies) [e.g. chrome, safari]
- Version of Next.js: 9.1.1

## Additional context

its probably also possible to debug that which just attaching events yourself on the browser history api

**Comment by seanggani** [comment]
Upvote this ticket, history api is called twice for me as well, wondering if anyone has any solution for this?

**Comment by kristijorgji** [comment]
Hi, am facing same issue as well but I think I have a solution for us (end of this post)

First page load is triggering twice router event and doubling the google analytics


My implementation of google analytics is very simple as documented in vercel examples + multiple posts (which I think have 1 line wrong)
Shortly:
In `_app.tsx` listen to router events and call gtag

```
    useEffect(() => {
        const handleRouteChange = (url: string) => {
            tracker.pageview(url);
        };
        router.events.on('routeChangeComplete', handleRouteChange);
        return () => {
            router.events.off('routeChangeComplete', handleRouteChange);
        };
    }, [router.events]);
```


In `_document.tsx` just add this script

```
<script
            dangerouslySetInnerHTML={{
                __html: `
                    window.dataLayer = window.dataLayer || [];
                    function gtag(){dataLayer.push(arguments);}
                    gtag('js', new Date());
                    gtag('config', '${Tracking.MEASUREMENT_ID}', {
                      page_path: window.location.pathname,
                    });
                 `,
            }}
        />
```

tracker.pageview is just doing this `gtag('config', this.measurementId, { page_path: path });`



I think we can easily solve this issue people by changing _document.tsx and removing this line (will test it out)

```
 gtag('config', '${Tracking.MEASUREMENT_ID}', {
                      page_path: window.location.pathname,
                    });
```
BECAUSE WE ALREADY TRACK ON ROUTER EVENT AS WELL

I think the documentation online is wrong and causing the double tracking issue, next router might be fine triggering the event routeChangeComplete for first route as well

**Reply by andirsun** [reply]
Why this solution is not in the official docs ? 
Big thanks bro you solve our analytics problems.

**Reply by kristijorgji** [reply]
I guess they forgot to add it bro, they have many docs to mantain. Yes that is worked fine for me deployed 2 weeks ago and all data is correct as compared to web before next.
Might write later a blog article with correct way to implement gtag in next as old blog posts around net use the double tracking

**Reply by josezm** [reply]
Hello @kristijorgji thank you for your reply. Do you have the link to your blog to see the Implementation ? Or some sample repository
